imagick/kernel_info.go

```diff
@@ -6,9 +6,145 @@ package imagick
 
 /*
 #include <magick/MagickCore.h>
+#include <magick/geometry.h>
+#include <magick/morphology.h>
 */
 import "C"
 
+//import "runtime"
+import "unsafe"
+
+// This struct represents the KernelInfo C API of ImageMagick
 type KernelInfo struct {
 	info *C.KernelInfo
 }
+
+func newKernelInfo(cki *C.KernelInfo) *KernelInfo {
+	ki := &KernelInfo{info: cki}
+	//TODO - understand and implement this.
+	// runtime.SetFinalizer(ki, Destroy)
+	//ki.IncreaseCount()
+	return ki
+}
+
+// Convert the current KernelInfo to an 2d-array of values. The values are either
+// a float64 if the element is used, or NaN if the element is not used by the kernel
+func (kernel_info *KernelInfo) ToArray() [][]float64 {
+	count := 0
+	values := [][]float64{}
+
+	for y := C.size_t(0); y < kernel_info.info.height; y++ {
+		row_values := make([]float64, kernel_info.info.width)
+		for x := C.size_t(0); x < kernel_info.info.width; x++ {
+			p2 := (*[1 << 10]C.double)(unsafe.Pointer(kernel_info.info.values))
+			row_values[x] = float64(p2[count])
+			count++
+		}
+		values = append(values, row_values)
+	}
+
+	return values
+}
+
+// Create a kernel from a builtin in kernel. See http://www.imagemagick.org/Usage/morphology/#kernel
+// for examples. Currently the 'rotation' symbols are not supported. Example:
+// kernel_info := imagick.AcquireKernelBuiltIn(imagick.KERNEL_RING, "2,1")
+func AcquireKernelBuiltIn(kernel_type KernelInfoType, kernelString string) *KernelInfo {
+	geometry_info := (*GeometryInfo)(C.malloc(C.size_t(unsafe.Sizeof(C.GeometryInfo{}))))
+	result := C.ParseGeometry(C.CString(kernelString), &geometry_info.gi)
+	var geometry_flags int = int(result)
+	FiddleWithGeometryInfo(kernel_type, geometry_flags, geometry_info)
+	kernel_info := C.AcquireKernelBuiltIn(C.KernelInfoType(kernel_type), &geometry_info.gi)
+
+	return newKernelInfo(kernel_info)
+}
+
+// ScaleKernelInfo() scales the given kernel list by the given amount, with or without
+// normalization of the sum of the kernel values (as per given flags). The exact behaviour
+// of this function depends on the normalization type being used please see
+// http://www.imagemagick.org/api/morphology.php#ScaleKernelInfo for details.
+// Flag should be one of:
+// imagick.KERNEL_NORMALIZE_NONE
+// imagick.KERNEL_NORMALIZE_VALUE
+// imagick.KERNEL_NORMALIZE_CORRELATE
+// imagick.KERNEL_NORMALIZE_PERCENT
+func (kernel_info *KernelInfo) Scale(scale float64, normalize_type KernelNormalizeType) {
```

Comment:
Go tends to prefer camelCase instead of snake_case. In some places you already have the camelCase, but have other places where it is snake. It also tends to prefer shorter names for receivers.. ie something like `k` or `ki` instead of `kernel_info`